File: src/vPoisson.m
function [Efield] = vPoisson(fs, grids, charge)
    
% compute charge density
rho = 0;
Ns = length(grids);
for s = 1:Ns
    rho = rho + charge(s)*sum(fs(:,:,s))*grids(s).dv;
end

Nx = grids(1).size(1);
kxn =  (mod((1:Nx)-ceil(Nx/2+1),Nx)- floor(Nx/2)) ;
Lx = grids(1).Lx;
rx = 2*pi*1i./Lx;
kx = kxn.*rx;

% laplacian is division -|k|^2
K2 = kx'.^2;
K2 ( abs(K2) < 1.0e-11 ) = 1;
% to avoid a division by zero, we set the zeroth wavenumber to one.
% this leaves it's respective Fourier coefficient unaltered, so the
% zero mode of Sk is conserved.dphi_dx_h = 1i*phi_fft.*kx(1,:); This way, Sk's zero mode implicitly
% defined the zero mode of the result
% Note that the zero mode is NOT uniquely defined: in a periodic
% setting, the solution of Laplace's (or Poisson's) equation is only
% defined up to a constant! You can freely overwrite the zero mode,
% therefore.
b = fft(rho); 
b(1) = 0;
phi_fft = - b' ./ K2; % solves second equation of vlassov poisson
dphi_dx_h = phi_fft.*kx';
    Efield = -reshape(ifft(dphi_dx_h, "symmetric"), 1, []);
    
end

File: python/src/vPoisson.py
import jax.numpy as jnp

def vPoisson(fs, grids, charge):
    """
    Solves the Poisson equation in Fourier space to compute the electric field.

    Args:
        fs (numpy array)      : Distribution function of shape (Nx, Nv, Ns).
        grids (list of dicts) : Grid information for each species.
        charge (list)         : Charge of each species.

    Returns:
        Efield (numpy array)  : Computed electric field.
    """
    
    ### Compute charge density for each species
    rho = jnp.zeros(grids[0]["Nx"])  # Initialize rho with zeros
    Ns = len(grids)                  # Number of species, but why not use params.Ns?
    
    for s in range(Ns):
        rho += charge[s] * jnp.sum(fs[:, :, s], axis=1) * grids[s]["dv"]

    # Get grid parameters
    Nx = grids[0]["Nx"]
    Lx = grids[0]["Lx"]
    
    # Define wave numbers
    kx = (2 * jnp.pi / Lx) * jnp.fft.fftshift(jnp.arange(-Nx//2, Nx//2))
    
    # Laplacian operator -|k|^2 (avoid division by zero)
    K2 = kx**2
    K2[0] = 1  # Set the first element to 1 to avoid division by zero
    
    # Solve Poisson equation in Fourier space
    b = jnp.fft.fft(rho)
    phi_fft = -b / K2
    phi_fft[0] = 0  # Set mean to zero
    
    # Compute electric field E = -dphi/dx
    dphi_dx_h = 1j * phi_fft * kx
    Efield = -jnp.real(jnp.fft.ifft(dphi_dx_h))  # Ensure real output
    
    return Efield